src/Authenticator/SessionAuthenticator.php

```diff
@@ -115,4 +117,68 @@ public function clearIdentity(ServerRequestInterface $request, ResponseInterface
             'response' => $response,
         ];
     }
+
+    /**
+     * Impersonates a user
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @param \ArrayAccess|array $impersonator
+     * @param \ArrayAccess|array $impersonated
+     * @return \ArrayAccess|array
+     */
+    public function impersonate(ServerRequestInterface $request, ResponseInterface $response, \ArrayAccess|array $impersonator, \ArrayAccess|array $impersonated): \ArrayAccess|array
+    {   $sessionKey = $this->getConfig('sessionKey');
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if ($session->check($impersonateSessionKey)) {
+            //throw new UnauthorizedException('You are already impersonating a user.');
+        }
+        $session->write($impersonateSessionKey, $impersonator);
+        $session->write($sessionKey, $impersonated);
+        $this->setConfig('identify', true);
+
+        return $impersonated;
+
+    }
+
+    /**
+     * Stops impersonation
+     *
+     * @param ServerRequestInterface $request
+     * @param ResponseInterface $response
+     * @return \ArrayAccess|array
+     */
+    public function stopImpersonate(ServerRequestInterface $request, ResponseInterface $response): \ArrayAccess|array
+    {
+        $sessionKey = $this->getConfig('sessionKey');
+        $impersonateSessionKey = $this->getConfig('impersonateSessionKey');
+        /** @var \Cake\Http\Session $session */
+        $session = $request->getAttribute('session');
+        if (!$session->check($impersonateSessionKey)) {
+           // throw new UnauthorizedException('You are not impersonating a user.');
+        }
```

Comment:
I think this is a reasonable error case to have.

Reply:
Removed //